src/test/scala/com/rbmhtechnology/calliope/GapDetectionSpec.scala

```diff
@@ -0,0 +1,225 @@
+
+      filterGapLess(3, events(3 to 5)) mustBe events(3 to 5)
+    }
+    "select a missing event within a consecutive gap if it appears within the persistence timeout" in {
+      filterGapLess(1, events(1 to 5)(gaps(3, 4))) mustBe events(1 to 2)
+
+      filterGapLess(3, events(3 to 5)(gaps(4))) mustBe events(3 to 3)
+
+      waitForPersistTimeout()
+
+      filterGapLess(4, events(4 to 5)(gaps(4))) mustBe events(5 to 5)
+    }
+    "select a missing event if it appears after the persistence timeout" in {
+      filterGapLess(1, events(1 to 5)(gaps(3))) mustBe events(1 to 2)
+
+      waitForPersistTimeout()
+
+      filterGapLess(3, events(3 to 5)) mustBe events(3 to 5)
+    }
+    "select all events after a previous after gaps were persisted" in {
+      implicit val g = gaps(4, 8)
+
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 3)
+
+      waitForPersistTimeout()
+
+      filterGapLess(4, events(4 to 9)) mustBe events(4 to 7)
+
+      waitForPersistTimeout()
+
+      filterGapLess(8, events(8 to 10)) mustBe events(8 to 10)
+
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 5)
+      filterGapLess(6, events(6 to 10)) mustBe events(6 to 10)
+    }
+    "select events up to a persisted gap even if gaps exist" in {
+      implicit val g = gaps(4, 8)
+
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 3)
+
+      waitForPersistTimeout()
+
+      filterGapLess(4, events(4 to 10)) mustBe events(4 to 8)
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 5)
+    }
+    "persist non-consecutive gaps" in {
+      implicit val g = gaps(4, 7)
+
+      filterGapLess(1, events(1 to 10)) mustBe events(1 to 3)
+
+      waitForPersistTimeout()
+
+      filterGapLess(4, events(4 to 10)) mustBe events(4 to 10)
+    }
+    "persist non-consecutive gaps for multiple selects" in {
+      implicit val g = gaps(3, 5, 8, 9)
+
+      filterGapLess(1, events(1 to 6)) mustBe events(1 to 2)
+      filterGapLess(1, events(1 to 10)) mustBe events(1 to 2)
+
+      waitForPersistTimeout()
+
+      filterGapLess(3, events(3 to 6)) mustBe events(3 to 6)
+      filterGapLess(3, events(3 to 10)) mustBe events(3 to 10)
+    }
+    "persist highest known gap" in {
+      implicit val g = gaps(4, 8)
+
+      filterGapLess(1, events(1 to 10)) mustBe events(1 to 3)
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 3)
+
+      waitForPersistTimeout()
+
+      filterGapLess(1, events(1 to 10)) mustBe events(1 to 10)
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 5)
+    }
+    "persist gaps as soon as the persistence-timeout has been reached even if newer gaps are found" in {
+      implicit val g = gaps(3, 4, 7, 11)
+
+      filterGapLess(1, events(1 to 10)) mustBe events(1 to 2)
+
+      sleep(timeout / 2)
+
+      filterGapLess(3, events(3 to 12)) mustBe empty
+
+      sleep(timeout * 3 / 4)
+
+      filterGapLess(3, events(3 to 12)) mustBe events(3 to 10)
+
+      waitForPersistTimeout()
```

Comment:
+1 but I guess the safe bet would be to use `timeout / 2` to ensure that the whole timeout has elapsed.